src/v/compression/async_stream_zstd.cc

```diff
@@ -0,0 +1,262 @@
+
+        return;
+    }
+
+    zstd_instance = std::make_unique<async_stream_zstd>(
+      decompression_size, default_contexts);
+}
+
+async_stream_zstd& async_stream_zstd_instance() {
+    if (unlikely(!zstd_instance)) {
+        initialize_async_stream_zstd(default_decompression_size);
+    }
+
+    return *zstd_instance;
+}
+
+async_stream_zstd::static_cctx::static_cctx(int compression_level) {
+    _c_buffer = ss::temporary_buffer<char>(ZSTD_CStreamOutSize());
+    auto c_workspace_size = ZSTD_estimateCStreamSize(compression_level);
+    // zstd requires alignment for the buffer below.
+    _c_workspace = ss::allocate_aligned_buffer<char>(c_workspace_size, 8);
+
+    vassert(
+      _c_workspace,
+      "Failed to allocate zstd workspace of size {}",
+      c_workspace_size);
+
+    _compress_ctx = ZSTD_initStaticCCtx(_c_workspace.get(), c_workspace_size);
+
+    throw_if_error(ZSTD_CCtx_setParameter(
+      _compress_ctx, ZSTD_c_compressionLevel, default_compression_level));
+}
+
+async_stream_zstd::static_cctx::static_cctx(static_cctx&& ctx) noexcept
+  : _compress_ctx(ctx._compress_ctx)
+  , _c_workspace(std::move(ctx._c_workspace))
+  , _c_buffer(std::move(ctx._c_buffer)) {
+    ctx._compress_ctx = nullptr;
+}
+
+async_stream_zstd::static_cctx&
+async_stream_zstd::static_cctx::operator=(static_cctx&& x) noexcept {
+    if (this != &x) {
+        _compress_ctx = x._compress_ctx;
+        _c_workspace = std::move(x._c_workspace);
+        _c_buffer = std::move(x._c_buffer);
+
+        x._compress_ctx = nullptr;
+    }
+
+    return *this;
+}
+
+async_stream_zstd::static_dctx::static_dctx(size_t decompression_size) {
+    _d_buffer = ss::temporary_buffer<char>(ZSTD_DStreamOutSize());
+    auto d_workspace_size = ZSTD_estimateDStreamSize(decompression_size);
+    // zstd requires alignment for the buffer below.
+    _d_workspace = ss::allocate_aligned_buffer<char>(d_workspace_size, 8);
+
+    vassert(
+      _d_workspace,
+      "Failed to allocate zstd workspace of size {}",
+      d_workspace_size);
+
+    _decompress_ctx = ZSTD_initStaticDCtx(_d_workspace.get(), d_workspace_size);
+}
+
+async_stream_zstd::static_dctx::static_dctx(static_dctx&& ctx) noexcept
+  : _decompress_ctx(ctx._decompress_ctx)
+  , _d_workspace(std::move(ctx._d_workspace))
+  , _d_buffer(std::move(ctx._d_buffer)) {
+    ctx._decompress_ctx = nullptr;
+}
+
+async_stream_zstd::static_dctx&
+async_stream_zstd::static_dctx::operator=(static_dctx&& x) noexcept {
+    if (this != &x) {
+        _decompress_ctx = x._decompress_ctx;
+        _d_workspace = std::move(x._d_workspace);
+        _d_buffer = std::move(x._d_buffer);
+
+        x._decompress_ctx = nullptr;
+    }
+
+    return *this;
+}
+
+async_stream_zstd::async_stream_zstd(
+  size_t decompression_size, int number_of_contexts) {
+    _decompression_size = decompression_size;
+
+    for (auto i = 0; i < number_of_contexts; i++) {
+        _compression_ctx_pool.release_object(
```

Comment:
It took me a while to understand this. This essentially is just initializing the pool with the N initial objects, right?

The call to release confused me. Maybe it's better to offer a constructor to the pool class which initializes the N objects (perhaps taking a lambda to create the objects, or constructor parameters), then the pool can offer only the "scoped" API, rather than the more error-prone release/get interfaces?

Since you are the only user currently I like the idea of offering only the scoped interface and if anyone needs an "unscoped" behavior in the future they can work on that part of the public API.

Reply:
Yeah, it's initializing the pool with N initial objects. The get/release interfaces are there if we want to built a timed_gc class on top of the object pool.